File: api/src/main/java/com/doerapispring/messaging/SpringApplicationEventsDomainEventPublisher.java
package com.doerapispring.messaging;

import com.doerapispring.domain.DomainEventPublisher;
import com.doerapispring.domain.DomainModel;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.PayloadApplicationEvent;
import org.springframework.stereotype.Component;

@Component
public class SpringApplicationEventsDomainEventPublisher implements DomainEventPublisher {
    private final ApplicationEventPublisher applicationEventPublisher;

    public SpringApplicationEventsDomainEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
        this.applicationEventPublisher = applicationEventPublisher;
    }

    @Override
    public void publish(DomainModel domainModel) {
        domainModel.getDomainEvents().forEach(domainEvent ->
            applicationEventPublisher.publishEvent(new PayloadApplicationEvent<>(domainEvent.getClass(), domainEvent)));
        domainModel.clearDomainEvents();
    }
}

File: api/src/main/scala/com/doerapispring/messaging/SpringApplicationEventsDomainEventPublisher.scala
package com.doerapispring.messaging

import com.doerapispring.domain.events.TodoListEvent
import com.doerapispring.domain.{DeprecatedTodoListModel, DomainEventPublisher, ListId, UserId}
import org.springframework.context.{ApplicationEventPublisher, PayloadApplicationEvent}
import org.springframework.stereotype.Component

@Component
class SpringApplicationEventsDomainEventPublisher(val applicationEventPublisher: ApplicationEventPublisher)
  extends DomainEventPublisher[DeprecatedTodoListModel, TodoListEvent, UserId, ListId] {

  override def publish(todoListModel: DeprecatedTodoListModel, todoListEvent: TodoListEvent, userId: UserId, listId: ListId): DeprecatedTodoListModel = {
    applicationEventPublisher.publishEvent(new  PayloadApplicationEvent[OwnedTodoListEvent](todoListModel, OwnedTodoListEvent(userId, listId, todoListEvent)))
    todoListModel
  }
}